src/Agent.Worker/Build/BuildDirectoryManager.cs

```diff
@@ -144,6 +145,78 @@ public sealed class BuildDirectoryManager : AgentService, IBuildDirectoryManager
+                    {
+                        throw new InvalidOperationException($"{path} contains invalid folder name characters.");
+                    }
+                    else if (path.Contains($"..{Path.DirectorySeparatorChar}") || path.Contains($"..{Path.AltDirectorySeparatorChar}"))
+                    {
+                        throw new InvalidOperationException($"{path} can not contains relative path.");
+                    }
+                    else if (Path.IsPathRooted(path))
+                    {
+                        throw new InvalidOperationException($"{path} can not be a rooted path.");
+                    }
+                    else
+                    {
+                        requiredRepositoryPath = Path.Combine(newConfig.SourcesDirectory, path);
+                    }
+                }
+                else
+                {
+                    // When repository doesn't has path set, default to sources directory 1/s
+                    requiredRepositoryPath = newConfig.SourcesDirectory;
+                }
+
+                Trace.Info($"Repository requires to be placed at '{requiredRepositoryPath}', current location is '{newConfig.RepositoryDirectory}'");
+                if (!string.Equals(newConfig.RepositoryDirectory, requiredRepositoryPath, IOUtil.FilePathStringComparison))
+                {
+                    executionContext.Output($"Repository is current at '{newConfig.RepositoryDirectory}', move to '{requiredRepositoryPath}'.");
+                    var count = 1;
+                    var staging = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), newConfig.BuildDirectory, $"_{count}");
+                    while (Directory.Exists(staging))
+                    {
+                        count++;
+                        staging = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), newConfig.BuildDirectory, $"_{count}");
+                    }
+
+                    var currentRepoPath = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), newConfig.RepositoryDirectory);
+                    var expectRepoPath = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), requiredRepositoryPath);
+                    try
+                    {
+                        executionContext.Debug($"Move existing repository '{currentRepoPath}' to '{expectRepoPath}' via staging directory '{staging}'.");
+                        IOUtil.MoveDirectory(currentRepoPath, expectRepoPath, staging, CancellationToken.None);
+                    }
+                    catch (Exception ex)
+                    {
+                        Trace.Error("Catch exception during repository move.");
+                        Trace.Error(ex);
+                        executionContext.Warning("Unable move existing repository to required location, try delete existing repository.");
+                        IOUtil.DeleteDirectory(currentRepoPath, CancellationToken.None);
```

Comment:
Should this line of code be like this instead?

```cs
IOUtil.DeleteDirectory(expectRepoPath, CancellationToken.None);
```

Otherwise, deleting `currentRepoPath` would leave you with no repository anywhere, right?

Reply:
if we can't perfectly move the folder into the right location, we better want to delete the current repository to avoid any unexpected repository corruption.